### ISSUE TITLE
Support for torch.fft

### ISSUE THREAD

**Issue by gorkemalgan** [pytorch/executorch#1362]
I have a custom model that uses `torch.fft.rfftn` and `torch.fft.irfftn`. I can successfully run `capture_pre_autograd_graph `and `export` (only with static sizes though). But, when I run `to_edge` I get the following error: 

Operator torch._ops.aten._fft_r2c.default is not Aten Canonical.

**Comment by gorkemalgan** [comment]
Thanks for the suggestion but now it throws the following error:

**Exception has occurred: SpecViolationError**
These operators are taking Tensor inputs with mismatched dtypes: defaultdict(<class 'dict'>, {<EdgeOpOverload: aten._fft_r2c.default>: schema = aten::_fft_r2c(Tensor self, int[] dim, int normalization, bool onesided) -> Tensor: {'self': torch.float32, '__ret_0': torch.complex64}, <EdgeOpOverload: aten.view_as_real_copy.default>: schema = aten::view_as_real_copy(Tensor self) -> Tensor: {'self': torch.complex64, '__ret_0': torch.float32}, <EdgeOpOverload: aten.complex.default>: schema = aten::complex(Tensor real, Tensor imag) -> Tensor: {'real': torch.float32, 'imag': torch.float32, '__ret_0': torch.complex64}, <EdgeOpOverload: aten._fft_c2r.default>: schema = aten::_fft_c2r(Tensor self, int[] dim, int normalization, SymInt last_dim_size) -> Tensor: {'self': torch.complex64, '__ret_0': torch.float32}})

**Comment by larryliu0820** [comment]
From step did you get this error?

**Comment by gorkemalgan** [comment]
`torch.fft.rfft` returns `torch.complex64` dtype tensor so as I understand there is no way to run it with ExecuTorch for now? Correct me if I am wrong, but I will need to wait till `torch.complex64` type is supported by ExecuTorch

**Comment by larryliu0820** [comment]
> `torch.fft.rfft` returns `torch.complex64` dtype tensor so as I understand there is no way to run it with ExecuTorch for now? Correct me if I am wrong, but I will need to wait till `torch.complex64` type is supported by ExecuTorch

Yep I don’t think we have complex number support in ExecuTorch runtime right now

**Comment by kimishpatel** [comment]
@guangy10 if we have triaging and followup meeting, I would like to follow up on this issue specifically.

@gorkemalgan note that there are a few issues here. 1) complex dtype not supported 2) implementation of *fft not available

**Comment by cccclai** [comment]
I think we already have the complex data type issue filed separately here: https://github.com/pytorch/executorch/issues/886

**Comment by cbilgin** [comment]
@gorkemalgan Closing because it's a duplicate. If this is a separate issue, please do open it again

**Comment by gorkemalgan** [comment]
This issue is not a duplicate of #886. In addition to complex data type support requested in #886 this issue requires Executorch implementation for torch.fft operator as well.

**Comment by cccclai** [comment]
> fft

I believe torch.fft is not part of core aten. If it is not, it'll decomposed to the core aten ops. Maybe @larryliu0820 or @SS-JIA can confirm

**Comment by SS-JIA** [comment]
With FFT related ops, we decided to defer decision for now. However, seeing that there is a use-case for them we will need to revisit whether those should be core. To my knowledge you can't really decompose fft.

However, even before considering adding fft ops, we need to think about how we're going to support complex data types. Is there a plan in place for this currently?

**Comment by alealv** [comment]
Hi, I'm also eager to have support of FFT functions.
In my case, I'm getting

```sh
    raise UnsupportedOperatorException(func)
torch._subclasses.fake_tensor.UnsupportedOperatorException: aten._fft_c2r.default

The above exception was the direct cause of the following exception:

...

    raise UnsupportedOperatorException(func)
RuntimeError: Failed running call_function <built-in method istft of type object at 0x7f3704de8a40>(*(FakeTensor(..., size=(s0, 513, 50), dtype=torch.complex64), 1024, 256, 1024, FakeTensor(..., size=(1024,))), **{'center': True}):
aten._fft_c2r.default

During handling of the above exception, another exception occurred:

....

    raise Unsupported(msg)
torch._dynamo.exc.Unsupported: unsupported operator: aten._fft_c2r.default (see https://docs.google.com/document/d/1GgvOe7C8_NVOMLOCwDaYV1mXXyHMXY7ExoewHqooxrs/edit#heading=h.64r4npvq0w0 for how to fix)

The above exception was the direct cause of the following exception:

....

    raise UserError(UserErrorType.DYNAMIC_CONTROL_FLOW, str(e)) from e
torch._dynamo.exc.UserError: speculate_subgraph: while introspecting cond, we were unable to trace function `tf_istft` into a single graph. This means that Dynamo was unable to prove safety for this API and will fall back to eager-mode PyTorch, which could lead to a slowdown. Scroll up for the stack trace of the initial exception. The reason was: unsupported operator: aten._fft_c2r.default (see https://docs.google.com/document/d/1GgvOe7C8_NVOMLOCwDaYV1mXXyHMXY7ExoewHqooxrs/edit#heading=h.64r4npvq0w0 for how to fix)
```

**Comment by olilarkin** [comment]
Is there any update on this? any branch? what would it take to implement aten::fft_rfft ?

thanks!

**Comment by swolchok** [comment]
I'm looking into implementing fft_r2c soon, but no promises yet on being able to do much with the resulting tensor beyond #8146.

**Comment by swolchok** [comment]
_fft_r2c is implemented, though it currently does not support onesided=false.
